File: android/src/main/java/com/visioncameraimagelabeler/VisionCameraImageLabelerPlugin.java
package com.visioncameraimagelabeler;

import android.annotation.SuppressLint;
import android.media.Image;

import androidx.camera.core.ImageProxy;

import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.label.ImageLabel;
import com.google.mlkit.vision.label.ImageLabeler;
import com.google.mlkit.vision.label.ImageLabeling;
import com.google.mlkit.vision.label.defaults.ImageLabelerOptions;
import com.mrousavy.camera.frameprocessor.FrameProcessorPlugin;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.concurrent.ExecutionException;

public class VisionCameraImageLabelerPlugin extends FrameProcessorPlugin {
  private final ImageLabeler labeler = ImageLabeling.getClient(ImageLabelerOptions.DEFAULT_OPTIONS);

  @Override
  public Object callback(ImageProxy frame, @NotNull Object[] params) {
    @SuppressLint("UnsafeOptInUsageError")
    Image mediaImage = frame.getImage();
    if (mediaImage != null) {
      InputImage image = InputImage.fromMediaImage(mediaImage, frame.getImageInfo().getRotationDegrees());
      Task<List<ImageLabel>> task = labeler.process(image);

      try {
        List<ImageLabel> labels = Tasks.await(task);

        WritableNativeArray array = new WritableNativeArray();
        for (ImageLabel label : labels) {
          WritableNativeMap map = new WritableNativeMap();
          map.putString("label", label.getText());
          map.putDouble("confidence", label.getConfidence());
          array.pushMap(map);
        }
        return array;
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
    return null;
  }

  VisionCameraImageLabelerPlugin() {
    super("labelImage");
  }
}

File: android/src/main/java/com/visioncameraimagelabeler/VisionCameraImageLabelerPlugin.kt
package com.visioncameraimagelabeler

import android.media.Image
import android.util.Log
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.label.ImageLabel
import com.google.mlkit.vision.label.ImageLabeling
import com.google.mlkit.vision.label.defaults.ImageLabelerOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin

private const val TAG = "ImageLabeler"

class VisionCameraImageLabelerPlugin() : FrameProcessorPlugin() {
  private val labeler = ImageLabeling.getClient(ImageLabelerOptions.DEFAULT_OPTIONS)

  override fun callback(frame: Frame, params: Map<String, Any>?): Any? {
    val mediaImage: Image? = frame.image
    if (mediaImage != null) {
      // TODO fix 0 param for proper orientation
      val image = InputImage.fromMediaImage(mediaImage, 0)
      val task = labeler.process(image)
      try {
        val labels: List<ImageLabel> = Tasks.await(task)
        // Use a mutable list of maps instead of WritableNativeArray
        val result = mutableListOf<Map<String, Any>>()
        for (label in labels) {
          result.add(mapOf("label" to label.text, "confidence" to label.confidence.toDouble()))
        }
        return result
      } catch (e: Exception) {
        Log.e(TAG, "Error processing image", e)
      }
    }
    return null
  }
}
